src/mscorlib/src/System/String.cs

```diff
@@ -2339,16 +2316,50 @@ public static int CompareOrdinal(String strA, String strB)
         //
         [Pure]
         [System.Security.SecuritySafeCritical]  // auto-generated
-        public static int CompareOrdinal(String strA, int indexA, String strB, int indexB, int length) {
-           if (strA == null || strB == null) {
-                if ((Object)strA==(Object)strB) { //they're both null;
+        public static int CompareOrdinal(String strA, int indexA, String strB, int indexB, int length)
+        {
+            if (strA == null || strB == null)
+            {
+                if (object.ReferenceEquals(strA, strB))
+                {
+                    // They're both null
                     return 0;
                 }
 
-                return (strA==null)? -1 : 1; //-1 if A is null, 1 if B is null.
+                return strA == null ? -1 : 1;
+            }
+
+            // COMPAT: Checking for nulls should become before the arguments are validated,
+            // but other optimizations which allow us to return early should come after.
+
+            if (length < 0)
+            {
+                // COMPAT: This was originally done in a call to CompareOrdinalEx which had the
```

Comment:
I do not think we need to preserve bug-for-bug compatibility here. If the actual argument is called `length`, the exception string should say length.